torch/csrc/autograd/cpp_hook.h

```diff
@@ -6,8 +6,7 @@
 namespace torch {
 namespace autograd {
 
-using hooks_list =
-    std::vector<std::function<at::TensorBase(const at::TensorBase&)>>;
+using hooks_list = std::vector<std::function<at::Tensor(const at::Tensor&)>>;
```

Comment:
Can we keep those as TensorBase?

Reply:
since TensorBase doesn't have is_same_size, we need to change it to Tensor here.

Reply:
also, the output of the hook needs to be passed to the check_single_result which is why the hook also can't be TensorBase

Reply:
https://github.com/pytorch/pytorch/blob/91eeb7726040f18513c920e19af1096952c42554/aten/src/ATen/native/cuda/Sort.cpp#L52
could this be used?

Reply:
Oh I actually forgot about OptionalTensorRefs. They actually won't work here due to the assertion that the tensors must be defined, but thanks to @zou3519 we have TensorRefs now. https://github.com/pytorch/pytorch/pull/105587.